### PR TITLE
NAS-126782 / 13.0 / PoolScrubService.__run: reduce memory usage during zpool history

### PR DESCRIPTION
**Pull request by rkojedzinszky**
This fixes memory fragmentation issues in python process, as it would allocate a big chunk of memory. This amount of memory is rarely allocated, and over time, the previously allocated block might get fragmented again.

### CODE REVIEW

File: src/middlewared/middlewared/plugins/pool.py [hunk 1]
```diff
@@ -4300,12 +4300,12 @@ async def __run(self, name, threshold):
         if pool['scan']['state'] == 'SCANNING':
             return False
 
-        history = (await run('zpool', 'history', name, encoding='utf-8')).stdout
-        for match in reversed(list(RE_HISTORY_ZPOOL_SCRUB.finditer(history))):
-            last_scrub = datetime.strptime(match.group(1), '%Y-%m-%d.%H:%M:%S')
-            break
+        last_scrub_str = (await run('sh', '-c', f'zpool history {name} | sed -r -n -e "s/{RE_HISTORY_ZPOOL_SCRUB_SED}.*/\\1/p" | tail -1', encoding='utf-8')).stdout.strip()
```
[the remark this answers is not in the document]
Comment (rkojedzinszky) [reply src/middlewared/middlewared/plugins/pool.py:4303]: According to https://openzfs.github.io/openzfs-docs/man/master/8/zpool-create.8.html#DESCRIPTION a pool name cannot contain spaces. Howewer, I totally agree to be safe in this case.

About the refactor: my intent was to give python as small data as possible. Comparing original output and just `grep`-d one:

```
root@nas-a[~]# zpool history main | wc
  337038 1442185 32653554
root@nas-a[~]# zpool history main | grep "zpool scrub" | wc
      85     340    3230
```

this also results in a very small data block, so I will rewrite as you suggested.

[the remark this answers is not in the document]
Comment (yocalebo) [reply src/middlewared/middlewared/plugins/pool.py:4303]: @rkojedzinszky the link you provided states ```The pool name must begin with a letter, and can only contain alphanumeric characters as well as the underscore, dash, colon, space, and period.``` Spaces are allowed in zpool names, unfortunately ☹️

[the remark this answers is not in the document]
Comment (rkojedzinszky) [reply src/middlewared/middlewared/plugins/pool.py]: Oh, I misread, sorry.

[the remark this answers is not in the document]
Comment (rkojedzinszky) [reply src/middlewared/middlewared/plugins/pool.py]: At first I was reading Oracle's documentation, and not assumed that openzfs had changed this, and missed the words there. (https://docs.oracle.com/cd/E23824_01/html/821-1448/gbcpt.html)

File: src/middlewared/middlewared/plugins/pool.py [hunk 2]
```diff
@@ -4300,16 +4300,16 @@ async def __run(self, name, threshold):
...
+        for match in reversed(list(RE_HISTORY_ZPOOL_SCRUB.finditer(last_scrubs))):
             last_scrub = datetime.strptime(match.group(1), '%Y-%m-%d.%H:%M:%S')
             break
         else:
             # creation time of the pool if no scrub was done
-            for match in RE_HISTORY_ZPOOL_CREATE.finditer(history):
-                last_scrub = datetime.strptime(match.group(1), '%Y-%m-%d.%H:%M:%S')
-                break
-            else:
+            creation = (await run('zfs', 'get', '-Hpo', 'value', 'creation', name, encoding='utf-8')).stdout
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/pool.py:4309]: Why run additional command? Previously we had only one command running, now we have two. Using `egrep "zpool scrub|zpool create"` will include both scrub and creation lines.

**Review by themylogin — CHANGES_REQUESTED** [review]

File: src/middlewared/middlewared/plugins/pool.py (hunk 2 above)
Comment (rkojedzinszky) [reply src/middlewared/middlewared/plugins/pool.py:4309]: Yes, we have 2 commands now. I think the code is very clear now. The first part runs zpool history, filtering for `zpool scrub` and parsing it for last scrub time. And only if not found, the second part is run. The second command directly queries the pool (dataset creation, which is the same) creation. There is no need for regular expressions or any kind of parsing. And, that command is run once ever during one pool's lifetime, as next time the first part will win, so it is not really an overhead. And I think the code is simpler.

**Review by themylogin — APPROVED** [review]
Thank you!

**Comment by rkojedzinszky** [comment]
Shall I squash the two commits into one?

**Comment by yocalebo** [comment]
> Shall I squash the two commits into one?

No, we can cherry-pick the commits no problem. We're going to backport this to 13.1 and to 23.10.2 and 24.04 since we're not releasing another 13.0 version at this time.

**Comment by yocalebo** [comment]
backport

**Pull request closed without merging**